c++/src/kj/common.h

```diff
@@ -1940,22 +1940,31 @@ class ArrayPtr: public DisallowConstCopyIfNotConst<T> {
   // Syntax sugar for invoking U::from.
   // Used to chain conversion calls rather than wrap with function.
 
-  inline void fill(T t) { 
+  inline void fill(T t) {
     // Fill the area by copying t over every element.
 
-    for (size_t i = 0; i < size_; i++) { ptr[i] = t; } 
+    for (size_t i = 0; i < size_; i++) { ptr[i] = t; }
     // All modern compilers are smart enough to optimize this loop for simple T's.
     // libc++ std::fill doesn't have memset specialization either.
   }
 
+  inline void fill(ArrayPtr<const T> other) {
+    // Fill the area by copying each element of other, in sequence, over every element.
+    if (other.size() == 0) return;
+    T* __restrict__ dst = begin();
+    for (size_t s = size_, i = 0; i < s; i++) {
+      dst[i] = other[i % other.size()];
```

Comment:
also paranoid hat on: can we trust compiler to hoist the size() or should we extract a immutable local?

Reply:
hmm... I think so but safer to extract I think

Reply:
For that matter the div operation is quite expensive, keeping a separate counter and resetting it to zero each time it reaches size would be a lot faster -- but entirely possible the compiler is smart enough to do that already.

Reply:
Done